### ISSUE TITLE
Does add_and_dot() imply taking the complex conjugate of one vector?

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#5747]
For #5746, I need to somehow extend GMRES to also work for complex-valued matrices. For this, I'm running into this function:
```
  /**
   * Performs a combined operation of a vector addition and a subsequent inner
   * product, returning the value of the inner product. In other words, the
   * result of this function is the same as if the user called
   * @code
   * this->add(a, V);
   * return_value = *this * W;
   * @endcode
   *
   * The reason this function exists is that this operation involves less
   * memory transfer than calling the two functions separately. This method
   * only needs to load three vectors, @p this, @p V, @p W, whereas calling
   * separate methods means to load the calling vector @p this twice. Since
   * most vector operations are memory transfer limited, this reduces the time
   * by 25\% (or 50\% if @p W equals @p this).
   */
  Number add_and_dot (const Number          a,
                      const Vector<Number> &V,
                      const Vector<Number> &W);
```

I guess the dot product implies that the function should really be taking the complex-conjugate of one of the two arguments and return something that's a real number. Would others agree? Should I make such a change?

**Comment by bangerth** [comment]
For reference, we have this:
```
  /**
   * Return the scalar product of two vectors.  The return type is the
   * underlying type of @p this vector, so the return type and the accuracy
   * with which it the result is computed depend on the order of the arguments
   * of this vector.
   *
   * For complex vectors, the scalar product is implemented as
   * $\left<v,w\right>=\sum_i v_i \bar{w_i}$.
   */
  template <typename Number2>
  Number operator * (const Vector<Number2> &V) const;
```

**Comment by masterleinad** [comment]
> I guess the dot product implies that the function should really be taking the complex-conjugate of one of the two arguments and return something that's a real number. Would others agree? Should I make such a change?

Yes, I agree that this makes sense.

**Comment by bangerth** [comment]
I've looked through the various classes and they all looked alright except for the CUDA vector (which is hopefully fixed in #5750). Specifically:
```
Vector  ok
VectorSpaceVector  abstract
TrilinosWrappers::Vector  TrilinosScalar is always real
PETScWrappers::VectorBase  ok
BlockVectorBase  ok
LinearAlgebra::Vector ok
LinearAlgebra::distributed::Vector ok
LinearAlgebra::distributed::BlockVector ok
CUDAVector fixed
LinearAlgebra::EpetraWrappers::Vector  TrilinosScalar is always real
```
I have a question for `VectorizedArray`, but for that I'll open a different issue.

**Comment by kronbichler** [comment]
I definitely agree that we should take the transpose of the vector `W`.
